sdk/core/System.ClientModel/tests/Options/PipelineMessageClassifierTests.cs

```diff
@@ -50,6 +53,253 @@ public void ClassifiesMultipleCodesAsNonErrors(int code, bool isError)
+        Assert.IsTrue(isError);
+
+        message.SetResponse(new MockPipelineResponse(403));
+        Assert.IsTrue(classifier.TryClassify(message, out isError));
+        Assert.IsFalse(isError);
+
+        message.SetResponse(new MockPipelineResponse(404));
+        Assert.IsTrue(classifier.TryClassify(message, out isError));
+        Assert.IsFalse(isError);
+
+        message.SetResponse(new MockPipelineResponse(500));
+        Assert.IsTrue(classifier.TryClassify(message, out isError));
+        Assert.IsTrue(isError);
     }
+
+    [Test]
+    public void CanComposeRetryClassifiers()
+    {
+        var last = PipelineMessageClassifier.Create(stackalloc ushort[] { 200, 201, 204 });
+
+        ChainingClassifier classifier = new ChainingClassifier(last);
+        classifier.AddClassifier(new RetriableStatusCodeClassifier(403, isRetriable: false));
+        classifier.AddClassifier(new RetriableStatusCodeClassifier(404, isRetriable: false));
+        classifier.AddClassifier(new RetriableStatusCodeClassifier(201, isRetriable: true));
+
+        MockPipelineMessage message = new();
+
+        message.SetResponse(new MockPipelineResponse(200));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out bool isRetriable));
+        Assert.IsFalse(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(201));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsTrue(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(204));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsFalse(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(304));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsFalse(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(403));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsFalse(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(404));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsFalse(isRetriable);
+
+        message.SetResponse(new MockPipelineResponse(500));
+        Assert.IsTrue(classifier.TryClassify(message, exception: default, out isRetriable));
+        Assert.IsTrue(isRetriable);
+    }
+
+    [Test]
+    public void CanComposeErrorAndRetryClassifiers()
+    {
+        var last = PipelineMessageClassifier.Create(stackalloc ushort[] { 200, 201, 204 });
+
+        ChainingClassifier classifier = new ChainingClassifier(last);
+        classifier.AddClassifier(new RetriableStatusCodeClassifier(429, isRetriable: false));
+        classifier.AddClassifier(new ErrorStatusCodeClassifier(404, isError: false));
+        classifier.AddClassifier(new ErrorStatusCodeClassifier(201, isError: true));
```

Comment:
With the new API, we can create a classifier that composes retry and error classifiers in a chain.